### PR TITLE
SSOP part generator

### LINKED ISSUE #24: Generate SSOP
Can be integrated into `generate_so.py`

### PR DESCRIPTION
**Pull request by dbrgn**
Generate SSOP parts based on JEDEC MO-150 and MO-152.

Based on #28, merge that first!

Fixes #24.

### CODE REVIEW

**Comment by rnestler** [comment]
Is there a PR already for the generated parts?

**Comment by dbrgn** [comment]
Yes, here: https://github.com/LibrePCB-Libraries/LibrePCB_Base.lplib/pull/35

File: generate_so.py [hunk 1]
```diff
@@ -514,4 +515,133 @@ def _make(dirpath: str) -> None:
         version='0.1',
         create_date='2019-07-21T12:15:54Z',
     )
+    _make('out/ssop')
+    _make('out/ssop/pkg')
```
Comment (rnestler) [inline generate_so.py]: Shouldn't `_make` create directories recursively?

File: generate_so.py [hunk 2]
```diff
@@ -514,4 +515,133 @@ def _make(dirpath: str) -> None:
...
+            #        N    e     D    E1   E    A
+
+            # 4.40mm body width
+            #   0.65mm pitch
+            SoConfig( 8,  0.65, 3.0, 4.4, 6.4, 2.0, 'AA'),
+            SoConfig(14,  0.65, 5.0, 4.4, 6.4, 2.0, 'AB-1'),
+            SoConfig(16,  0.65, 5.0, 4.4, 6.4, 2.0, 'AB'),
+            SoConfig(20,  0.65, 6.5, 4.4, 6.4, 2.0, 'AC'),
+            SoConfig(24,  0.65, 7.8, 4.4, 6.4, 2.0, 'AD'),
+            SoConfig(28,  0.65, 9.7, 4.4, 6.4, 2.0, 'AE'),
```
Comment (rnestler) [inline generate_so.py]: Do you have them explicitly for some reasons or would it be easier to create it with a list comprehension or similar?

File: generate_so.py [hunk 3]
```diff
@@ -514,4 +515,133 @@ def _make(dirpath: str) -> None:
...
+            SoConfig(38, 0.65, 12.6, 5.3, 7.8, 2.0, 'AK'),
+        ],
+        lead_width_lookup={
+            0.65: 0.38,
+        },
+        lead_contact_length=0.75,
+        pkgcat='3627bf02-2e6e-4d68-9ada-743fa69a4f8c',
+        keywords='so,sop,ssop,small outline package,smd,jedec,mo-150',
+        version='0.1',
+        create_date='2019-07-21T12:55:20Z',
```
Comment (rnestler) [inline generate_so.py]: Just a small detail, but should we use a more current date?

File: generate_so.py [hunk 4]
```diff
@@ -514,4 +515,133 @@ def _make(dirpath: str) -> None:
...
+        lead_width_lookup={
+            0.65: 0.30,
+            0.50: 0.27,
+            0.40: 0.23,
+        },
+        lead_contact_length=0.6,
+        pkgcat='3627bf02-2e6e-4d68-9ada-743fa69a4f8c',
+        keywords='so,sop,ssop,small outline package,smd,jedec,mo-152',
+        version='0.1',
+        create_date='2019-07-21T12:55:20Z',
```
Comment (rnestler) [inline generate_so.py]: Same here: Should we use a more current date?

File: generate_so.py (hunk 2 above)
Comment (dbrgn) [reply generate_so.py]: It's easier to read that way. Almost an 1:1 copy from the JEDEC standard.

File: generate_so.py (hunk 3 above)
Comment (dbrgn) [reply generate_so.py]: Doesn't really matter, it's the date when I started working on those parts.

As far as I know @ubruhin even thought about removing the date completely. It doesn't have any function.

File: generate_so.py (hunk 1 above)
Comment (dbrgn) [reply generate_so.py]: True, since it uses `os.makedirs` under the hood.

File: generate_so.py (hunk 2 above)
Comment (rnestler) [reply generate_so.py]: Sounds reasonable.

File: generate_so.py (hunk 3 above)
Comment (rnestler) [reply generate_so.py]: Yeah, seeing a last updated date (gut log) may be interesting to see if the library is still maintained, but created date is a bit moot.

File: uuid_cache_so.csv [hunk 5]
```diff
@@ -13023,6 +13023,7506 @@ pkg-soic127p762x270-8-text-name-density~c,6706e38b-e92c-4458-908e-b83edcd2dcd3
...
+pkg-ssop24p40_500x640x200l100x23-pad-7,f089997b-4c2a-4dc9-9236-1f14954d684e
+pkg-ssop24p40_500x640x200l100x23-pad-8,8320c65d-f90e-45f3-989a-715414a807d3
+pkg-ssop24p40_500x640x200l100x23-pad-9,3949373f-a86a-4e4c-9438-39e2d85a2960
+pkg-ssop24p40_500x640x200l100x23-pkg,dfcd74f6-c903-43eb-b6ee-63d1dbccb6c7
+pkg-ssop24p40_500x640x200l100x23-polygon-courtyard-density~a,b7734861-9996-41ac-b36f-c3b9e239463b
+pkg-ssop24p40_500x640x200l100x23-polygon-courtyard-density~b,e3fd83db-9af3-4d4f-9a76-48e1e43bf94c
+pkg-ssop24p40_500x640x200l100x23-polygon-courtyard-density~c,6c3ac7b3-6486-4c2e-bac5-599d8e48bfdb
+pkg-ssop24p40_500x640x200l100x23-polygon-outline-density~a,1f8d0783-f7e4-4493-ad6f-ef281a2e426a
+pkg-ssop24p40_500x640x200l100x23-polygon-outline-density~b,56588f58-5a2d-4150-90a3-0e2eb5ca140e
+pkg-ssop
```
Comment (ubruhin) [inline uuid_cache_so.csv]:
```suggestion
pkg-ssop28p65_1020x780x200l125x38-footprint-density~b,aacb0431-7029-474c-b88e-452820371aa5
```

**Comment by ubruhin** [comment]
Arghh, somehow the suggested changes look wrong:
grafik

There is no line with only `pkg-ssop`, no idea whether applying the suggestion will do the correct change :sob:

**Comment by dbrgn** [comment]
No problem, I'll apply them manually.

File: uuid_cache_so.csv (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbrgn) [reply uuid_cache_so.csv]: Applied in 9cf6794725226b9ca918ddbcd5a808186f259e3f

Comment (dbrgn) [reply uuid_cache_so.csv]: Applied manually in 9cf6794725226b9ca918ddbcd5a808186f259e3f

Comment (dbrgn) [reply uuid_cache_so.csv]: Applied in 733a3901a9865c59b67b076b7b3e49bf6cc77222

**Comment by dbrgn** [comment]
@ubruhin I updated https://github.com/LibrePCB-Libraries/LibrePCB_Base.lplib/pull/35 with your changes.

**Comment by dbrgn** [comment]
Oh, and regarding this question:

> Beside this: Are pad positions rounded to 2 decimal places somewhere? I ask because in the packages with 0.65mm pad spacing and even number of pads per side (e.g. SSOP28), pads have an Y offset of +/-0.33mm instead of +/-0.325mm. Thus the two pads in the middle have a spacing of 0.66mm instead of 0.65mm.

Hm, you're right. In general (`format_float(...)`) we round to 3 decimal places, but in the `get_y` function in `generate_so.py` we round to 2 decimal places. I'll adjust that.

**Comment by dbrgn** [comment]
Changed in ef6a98a576c05d311353c00c1be3e6d9bfe5c78b. I pushed fixup commits to https://github.com/LibrePCB-Libraries/LibrePCB_Base.lplib/pull/35. The other affected parts will have to be updated in a separate PR.

File: uuid_cache_so.csv (hunk 5 above)
Comment (ubruhin) [inline uuid_cache_so.csv]: Oops, I missed to change the most important one of the UUIDs, sorry! :sob:

```suggestion
pkg-ssop28p65_1020x780x200l125x38-pkg,2bdb4572-0418-4f98-8cce-1e128f01afd4
```

Comment (dbrgn) [reply uuid_cache_so.csv]: Fixed.

**Comment by dbrgn** [comment]
Library PR is merged!

**Pull request merged**